lib/models/room.js

```diff
@@ -438,245 +456,94 @@ Room.prototype.getCanonicalAlias = function() {
-    // timeline we ended up on.
-    if (lastEventWasNew || !didUpdate) {
-        timeline.setPaginationToken(paginationToken, direction);
-    }
+    return timelineSet;
 };
 
 /**
- * Add event to the given timeline, and emit Room.timeline. Assumes
- * we have already checked we don't know about this event.
+ * Forget the timelineSet for this room with the given filter
  *
- * Will fire "Room.timeline" for each event added.
- *
- * @param {MatrixEvent} event
- * @param {EventTimeline} timeline
- * @param {boolean} toStartOfTimeline
- *
- * @fires module:client~MatrixClient#event:"Room.timeline"
- *
- * @private
+ * @param {Filter} filter  the filter whose timelineSet is to be forgotten
  */
-Room.prototype._addEventToTimeline = function(event, timeline, toStartOfTimeline) {
-    var eventId = event.getId();
-    timeline.addEvent(event, toStartOfTimeline);
-    this._eventIdToTimeline[eventId] = timeline;
-
-    var data = {
-        timeline: timeline,
-        liveEvent: !toStartOfTimeline && timeline == this._liveTimeline,
-    };
-    this.emit("Room.timeline", event, this, Boolean(toStartOfTimeline), false, data);
+Room.prototype.removeFilteredTimelineSet = function(filter) {
+    var timelineSet = this._filteredTimelineSets[filter.filterId];
+    delete this._filteredTimelineSets[filter.filterId];
+    var i = this._timelineSets.indexOf(timelineSet);
+    if (i > -1) {
+        this._timelineSets.splice(i, 1);
+    }
 };
 
-
 /**
- * Add an event to the end of this room's live timeline. Will fire
- * "Room.timeline"..
+ * Add an event to the end of this room's live timelines. Will fire
+ * "Room.timeline".
  *
  * @param {MatrixEvent} event Event to be added
  * @param {string?} duplicateStrategy 'ignore' or 'replace'
  * @fires module:client~MatrixClient#event:"Room.timeline"
  * @private
  */
 Room.prototype._addLiveEvent = function(event, duplicateStrategy) {
+    var i;
     if (event.getType() === "m.room.redaction") {
         var redactId = event.event.redacts;
 
-        // if we know about this event, redact its contents now.
-        var redactedEvent = this.findEventById(redactId);
-        if (redactedEvent) {
-            redactedEvent.makeRedacted(event);
-            this.emit("Room.redaction", event, this);
-
-            // TODO: we stash user displaynames (among other things) in
-            // RoomMember objects which are then attached to other events
-            // (in the sender and target fields). We should get those
-            // RoomMember objects to update themselves when the events that
-            // they are based on are changed.
+        for (i = 0; i < this._timelineSets.length; i++) {
+            var timelineSet = this._timelineSets[i];
+            // if we know about this event, redact its contents now.
+            var redactedEvent = timelineSet.findEventById(redactId);
+            if (redactedEvent) {
+                redactedEvent.makeRedacted(event);
```

Comment:
You're sharing the `MatrixEvent` between the timelineSets, so you probably only want to do this once (for the unfiltered timeline, presumably). Not sure that doing it multiple times is going to hurt anything, but it's misleading at best.